cloudslang-runtime/src/main/java/io/cloudslang/lang/runtime/steps/ActionExecutionData.java

```diff
@@ -193,9 +205,44 @@ public void doAction(@Param(EXECUTION_RUNTIME_SERVICES) ExecutionRuntimeServices
         if (returnMap == null) {
             throw new RuntimeException("Action method did not return Map<String,String>");
         }
+
+        if (hasException(returnMap)) {
+            logException(returnMap);
+        }
+
         return handleSensitiveValues(returnMap, currentContext);
     }
 
+    private boolean hasException(Map<String, Serializable> returnMap) {
+        return returnMap.containsKey(RETURN_CODE) && !"0".equals(returnMap.get(RETURN_CODE)) &&
+            returnMap.containsKey(EXCEPTION);
+    }
+
+    private void logException(Map<String, Serializable> returnMap) {
+
+        String stacktrace = (String) returnMap.get(EXCEPTION);
+
+        if (REMOVE_MESSAGE_FROM_LOGGED_EX) {
+            StringWriter writer = new StringWriter();
+            final PrintWriter printWriter = new PrintWriter(writer);
+
+            BufferedReader reader = new BufferedReader(new StringReader(stacktrace));
+
+            reader.lines().forEach(line -> {
```

Comment:
lambdas should be stateless, should not touch variables from ouside like printWriter, consider having a standard for